src/pam_hbac_ldap_compat.c

```diff
@@ -0,0 +1,284 @@
+    new_rdn = calloc(2, sizeof(LDAPAVA *));
+    new_rdn[0] = rdn2ava(str_rdn);
+    if (new_rdn[0] == NULL) {
+        free(new_rdn);
+        return ENOMEM;
+    }
+    new_rdn[1] = NULL;
+
+    *_rdn = new_rdn;
+    return 0;
+}
+#endif
+
+int ph_str2dn(const char *str, LDAPDN *dn)
+{
+#if defined(HAVE_LDAP_STR2DN) && !defined(COMPAT_LDAP_UNIT_TESTS)
+    return ldap_str2dn(str, dn, LDAP_DN_FORMAT_LDAPV3);
+#else
+    char **str_rdn_list;
+    size_t n_rdns;
+    size_t i;
+    LDAPDN new_dn;
+    LDAPRDN rdn;
+    int ret;
+
+    if (str == NULL || dn == NULL) {
+        return EINVAL;
+    }
+
+    str_rdn_list = ldap_explode_dn(str, 0);
+    if (str_rdn_list == NULL) {
+        return ENOMEM;
+    }
+
+    for (n_rdns = 0; str_rdn_list[n_rdns]; n_rdns++);
+
+    new_dn = calloc(n_rdns + 1, sizeof(LDAPRDN));
+    if (new_dn == NULL) {
+        ldap_value_free(str_rdn_list);
+        return ENOMEM;
+    }
+
+    for (i = 0; str_rdn_list[i]; i++) {
+        ret = str2rdn(str_rdn_list[i], &rdn);
+        if (ret != 0) {
+            ldap_value_free(str_rdn_list);
+            ph_ldap_dnfree(new_dn);
+            return ret;
+        }
+
+        new_dn[i] = rdn;
+    }
+
+    ldap_value_free(str_rdn_list);
+    *dn = new_dn;
+    return LDAP_SUCCESS;
+#endif
+}
+
+#if !defined(HAVE_LDAP_INITIALIZE) || defined(COMPAT_LDAP_UNIT_TESTS)
+static ssize_t uri_proto_prefix_len(const char *uri)
+{
+
+    if (strncasecmp(uri, "ldap://", sizeof("ldap://")-1) == 0) {
+        /* The server is IPA and we don't support anything else */
+        return sizeof("ldap://")-1;
+    }
+
+#ifdef HAVE_LDAPSSL_CLIENT_INIT
+    /* Older MozLDAP/SunLDAP libraries don't support StartTLS but only
+     * SSL.
+     */
+    if (strncasecmp(uri, "ldaps://", sizeof("ldaps://")-1) == 0) {
+        /* The server is IPA and we don't support anything else */
+        return sizeof("ldaps://")-1;
+    }
+#endif
+
+    return -1;
+}
+#endif
+
+/* ldap_initialize version for systems that only have ldap_init.
+ * More or less stolen from nss-pam-ldapd
+ */
+int ph_ldap_initialize(LDAP **ld, const char *uri)
+{
+#if defined(HAVE_LDAP_INITIALIZE) && !defined(COMPAT_LDAP_UNIT_TESTS)
+    return ldap_initialize(ld, uri);
+#else
+    char hostname[HOST_NAME_MAX+1];
+    size_t hostlen;
+    ssize_t uri_prefix;
+
+    if (ld == NULL || uri == NULL) {
+        return EINVAL;
+    }
+
+    uri_prefix = uri_proto_prefix_len(uri);
+    if (uri_prefix == -1) {
+        return LDAP_INVALID_SYNTAX;
+    }
+
+    strncpy(hostname, uri + uri_prefix, sizeof(hostname));
+    hostname[HOST_NAME_MAX] = '\0';
+
+    hostlen = strlen(hostname);
+    if (hostlen == 0) {
+        return LDAP_INVALID_SYNTAX;
+    }
+
+    *ld = ldap_init(hostname, LDAPS_PORT);
+    if (*ld == NULL) {
+        return LDAP_OPERATIONS_ERROR;
+    }
+
+    return LDAP_SUCCESS;
+#endif
+}
+
+#if !defined(HAVE_LDAP_DNFREE) || defined(COMPAT_LDAP_UNIT_TESTS)
+static void free_ava(LDAPAVA *ava)
+{
+    free(ava->la_attr.bv_val);
```

Comment:
It seems that currently NULL cannot be passed as ava parameter...should we be paranoid?